Add TLS encryption support to MQTT

mtreinish opened the pull request:
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

- notification/mqtt

##### ANSIBLE VERSION
```
ansible 2.1.1.0
```

##### SUMMARY
This commit adds module settings for configuring TLS encryption on the
mqtt notification module. Previously there was no way to configure
sending the messages encrypted to mqtt. This is necessary in some mqtt
environments, especially given the basic auth support without encryption
enabled sends credentials in clear text.

gregdek commented:
Thanks @mtreinish. To the current maintainers, @jpmens please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

pabelanger commented:
+1

mtreinish commented:
The latest revision just updates the version_added tag in the docs for the new options. (per what travis pointed out)

jpmens commented:
shipit

gregdek commented:
Thanks again to @mtreinish for this PR, and thanks @jpmens for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

resmo commented on notification/mqtt.py:
```diff
@@ -75,6 +75,36 @@
         retained message immediately.
     required: false
     default: False
+  ca_certs:
```
it arg_spec, it is defined as `ca_cert`

resmo commented on notification/mqtt.py:
```diff
@@ -75,6 +75,36 @@
+    required: False
+    default: None
+    version_added: 2.2
+  keyfile:
```
in arg_spec it is defined as `key_file`

resmo commented on notification/mqtt.py:
```diff
@@ -121,6 +151,11 @@ def main():
             retain = dict(default=False, type='bool'),
             username = dict(default = None),
             password = dict(default = None, no_log=True),
+            ca_cert = dict(default = None),
```
please add `type='path'` to the arg_spec

resmo commented on notification/mqtt.py:
```diff
@@ -121,6 +151,11 @@ def main():
             username = dict(default = None),
             password = dict(default = None, no_log=True),
+            ca_cert = dict(default = None),
+            certfile = dict(default = None),
```
please add `type='path'` to the arg_spec

resmo commented on notification/mqtt.py:
```diff
@@ -121,6 +151,11 @@ def main():
             password = dict(default = None, no_log=True),
+            ca_cert = dict(default = None),
+            certfile = dict(default = None),
+            key_file = dict(default = None),
```
please add `type='path'` to the arg_spec

resmo commented:
I made a few inline comments and I am missing docs for 2 arg_specs:

``` diff
+            tls_version = dict(default = None),
+            ciphers = dict(default = None),
```

mtreinish commented:
I've updated the commit per your inline comments. Also the 2 arguments with missing docs weren't actually ever used any where (I must have just sloppily copy and pasted them in from the paho arg list) So I just removed them.

mtreinish commented:
I just noticed some other places I was sloppy with ca_cert vs ca_certs, key_file vs keyfile, etc. The latest rev should have that all cleaned up now.

gregdek commented:
@mtreinish issue at this point is that we're in feature freeze for the 2.2 release, which means it's likely that this feature won't make it in until 2.3.

mtreinish commented:
@gregdek ah, ok. Thanks for the details.

gregdek commented:
Thanks @mtreinish for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

mtreinish commented:
ready_for_review

mtreinish commented:
The latest rev bumps the version added per the CI check

mtreinish commented:
ready_for_review

gregdek commented:
Thanks @mtreinish. To the current maintainers, @jpmens please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

jpmens commented:
shipit

gregdek commented:
Thanks again to @mtreinish for this PR, and thanks @jpmens for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]
